ansible/roles/dev/tasks/main.yml

```diff
@@ -17,20 +17,39 @@
   notify: restart sshd
 
 - name: Install python-virtualenvwrapper
-  package: name=python-virtualenvwrapper state=latest
+  package:
+    name: "{{ item }}"
+    state: latest
+  with_items:
+    # FIXME(milan): a dependency bug?
+    - which
+    - python3-virtualenvwrapper
+    # FIXME(milan): the python2 version is required for mkvenv as mkvenv
+    # doesn't seem to respect the VIRTUALENVWRAPPER_PYTHON variable
+    - python2-virtualenvwrapper
   when: pulp_venv is defined
 
 - block:
+  - name: Upgrade pip in the pulp venv
+    # NOTE(milan): besides avoiding possible issues this reduces debugging error
+    # logs by pip ranting about being out of date all the time
+    pip:
+      name: pip
```

Comment:
should this be pip3?

Reply:
I think it needn't: `pip` is being updated in the `pulp` virtual environment where it defaults to `pip3`:

```
(pulp) [vagrant@pulp3 pulp]$ which pip
~/.virtualenvs/pulp/bin/pip
(pulp) [vagrant@pulp3 pulp]$ ~/.virtualenvs/pulp/bin/pip --version
pip 9.0.1 from /home/vagrant/.virtualenvs/pulp/lib64/python3.6/site-packages (python 3.6)
(pulp) [vagrant@pulp3 pulp]$
```

Reply:
No errors but every pip command in the virtual env complains about pip being out of date (when run in verbose mode).
Unless I missed some occurrence this is the first command issued in the pulp venv (creating it as a side effect).

Reply:
> Unless I missed some occurrence this is the first command issued in the pulp venv (creating it as a side effect).

Yes, this is the first task created after the pulp venv is created. The problem is, it will execute every subsequent time the playbook is run. A good Ansible playbook is idempotent, and this task harms the idempotency of any playbook using it.

Consider upgrading pip when the venv is first created. Something like:

```yaml
- name: Create venv
  ...
  register: result

- name: Update pip
  ...
  when: result|changed
```

Reply:
Any better way to create a venv than to cmd/shell?

Reply:
The code here is a good example of how to create a Python 3 venv without using the `command` or `shell` modules.

Reply:
you mean to upgrade pip having installed e.g `pulp.common`?

Reply:
No, I'm not talking about upgrading pip, or installing `pulp.common`. I'm answering the question above, which is how to create a venv without using the `command` or `shell` modules. Do you have a question about upgrading pip, or installing `pulp.common`?

Reply:
Sorry for not being entirely comprehensible; 
I mean what would you suggest to put instead of the ellipsis here:

```yaml
- name: Create venv
   ...
  register: result
```
...to actually create the venv

Reply:
Ahh, I see now. Here's a sample playbook that creates a Python 3 venv:

```yaml
- hosts: all
  tasks:
    - name: Install requests
      pip:
        name: requests
        state: present
        virtualenv: "/home/dparalen/.venvs/test-env"
        virtualenv_command: /usr/bin/python3 -m venv
```

A couple points of interest:

* The `virtualenv` argument must be associated with an absolute path. Relative paths will not work.
* In this example, `/home/dparalen/.venvs` doesn't need to exist. More generally, the directory containing the venv being created doesn't need to exist. Intermediary directories will be created.
* The only dependency is Python 3.3+. Packages like `virtualenv`, `python2-virtualenv`, `python3-virtualenv`, and the like do not need to be installed.
* If you'd like to test this playbook, the following command should work, assuming this playbook is in a file named `site.yml`: `ansible-playbook site.yml -i localhost, --connection local`.

Reply:
By the way, if you'd like to reference a virtuenv without explicitly hard-coding absolute paths into a playbook, consider a strategy like this:

```yaml
- hosts: all
  tasks:
    - name: Get information about dparalen
      getent:
        database: passwd
        key: dparalen
        split: ':'

    - name: "Set variable referencing venv"
      set_fact:
        my_venv: "{{ getent_passwd['dparalen'][4] }}/.venvs/my-venv"

    - name: Install requests
      pip:
        name: requests
        state: present
        virtualenv: "{{ my_venv }}"
        virtualenv_command: /usr/bin/python3 -m venv
```

Make sure to prefix the variable with the name of the role it's being created in. This is because variables continue to exist after their containing role has finished executing, and they can have unexpected impacts on other roles. In this case, `my_venv` should be prefixed with something.

Reply:
Thanks for the guide!
I've done some more checking and it turns out F26 already ships latest pip (9.1.z) so this isn't needed.

Cheers,
milan